### PR TITLE
no coalesced keys + key separator

### PR DESCRIPTION
**Pull request by joanvr**
## What's changed?
Do not coalesce keys by default. The current behavior was to always coalesce the keys into a single key separated by dots when possible, but it's not actually true that for yaml they are equivalent. So instead of:
```
a.b.c: true
```
we will have:
```
a:
  b:
    c: true
```
For scenarios when keys with dots are required, added an extra configurable parameter to change the key separator. So we can support having
```
a:
  b.c: true
```
if that's the intended behavior.

## What's your motivation?
This recipe has a lot of tests and involved in a lot of issues (some of them still open) regarding coalescing keys (or not coalescing them), and this has implications when later on we also use the MergeYaml visitor to merge results, since MergeYaml does not consider to be the same
```
a: 
  b: true
```
than
```
a.b: true
```
because in yaml specification, it is not... Just in some scenarios and frameworks that uses yaml for configuration/properties files, this happens to be the same.

## Anything in particular you'd like reviewers to focus on?
I'm worried that changing the recipe expectations/api, this could cause unwanted behavior in people who is using this recipe, although, based on the tests and open issues, the new behavior is the expected/wanted by most users. However, we are only using this recipe in Spring [here](https://github.com/openrewrite/rewrite-spring/blob/f2e9faffdabec06550c1eeacc4da185dde30f1c4/src/main/java/org/openrewrite/java/spring/ChangeSpringPropertyKey.java#L67) and for people that is using this recipe in a pinned version, it might not even be a problem

## Anyone you would like to review specifically?
@timtebeek @shanman190 

## Any additional context
I think that this requires more discussion/agreement, maybe we can use this PR to come to an agreement.
Also, "fixed" all the existing tests to pass, but I have the feeling that a lot of them were added due to "issues" in coalesced keys that don't have sense anymore, that could be remove because they are redundant...

### Checklist
- [x] I've added unit tests to cover both positive and negative cases
- [x] I've updated the documentation (if applicable)

### CODE REVIEW

**Comment by shanman190** [comment]
I think overall the change looks fine. One alternative to note is the possible use of the jsonpath style. With the Json path style properties, they opt to quote properties with dots in them that should not be separated (ex: `a.b."c.d".e.f`). Either the current implementation or a jsonpath style seems like they would get the job done accurately from my perspective.

**Comment by joanvr** [comment]
> I think overall the change looks fine. One alternative to note is the possible use of the jsonpath style. With the Json path style properties, they opt to quote properties with dots in them that should not be separated (ex: `a.b."c.d".e.f`). Either the current implementation or a jsonpath style seems like they would get the job done accurately from my perspective.

I also thought of this approach, to use quotes or ticks to specify keys that cannot be separated, the problem is that (I believe) even all those characters are valid characters for a key according to the spec, although it's probably super uncommon...

**Comment by joanvr** [comment]
Some related issues that would fix with this:
https://github.com/openrewrite/rewrite/issues/3426
https://github.com/openrewrite/rewrite/issues/1249
https://github.com/openrewrite/rewrite/issues/1841
https://github.com/openrewrite/rewrite/issues/1873
https://github.com/openrewrite/rewrite/issues/1114
https://github.com/openrewrite/rewrite-spring/issues/189

File: rewrite-yaml/src/test/java/org/openrewrite/yaml/ChangePropertyKeyTest.java [hunk 1]
```diff
@@ -468,13 +471,16 @@ void relocatesPropertyWithSamePrefix() {
...
+                        interval: 1
               """
           )
         );
     }
 
     @Nested
+    @Disabled
     @Issue("https://github.com/openrewrite/rewrite-spring/issues/189")
     class WhenOldPropertyKeyIsPrefixOfDotSeparatedKeyTest implements RewriteTest {
```
Comment (timtebeek) [inline rewrite-yaml/src/test/java/org/openrewrite/yaml/ChangePropertyKeyTest.java:485]: What should we do about these test cases? Looked to be added for https://github.com/openrewrite/rewrite-spring/issues/189 by @nmck257 
Only when I add an explicit non null, non default separate does the first one pass again.

Does that mean there's a risk we break things for people that were using the recipe as it was before a merge?

Comment (nmck257) [reply rewrite-yaml/src/test/java/org/openrewrite/yaml/ChangePropertyKeyTest.java:485]: Thanks for the tag -- hadn't seen this PR.
Conceptually, I'm always going to be in favor of these yml utilities better-preserving existing style in the source code when it comes to dot separation vs indents. Will review these changes more specifically soon

Comment (nmck257) [reply rewrite-yaml/src/test/java/org/openrewrite/yaml/ChangePropertyKeyTest.java:485]: I believe `WhenOldPropertyKeyIsPrefixOfDotSeparatedKeyTest` is mostly here to contrast with the `ExceptTest` below it. So in that sense, the exact output of these particular cases is not super-important, because `rewrite-spring` is now doing what's shown in `ExceptTest` for this specific property migration, not what's shown in `WhenOldPropertyKeyIsPrefixOfDotSeparatedKeyTest`.

**Comment by nmck257** [comment]
have we tested a snapshot of this with `rewrite-spring`, and `ChangeSpringPropertyKeyTest`?

It looks like the strategy for the tests in this PR has been to adjust the recipe arguments to recreate the exact existing `before/after`. That feels like the wrong choice.
We know in the short term that existing usages of this recipe will bring their existing arguments, and we expect in the long term that most usages of this recipe would still choose to use the default value for this new argument. Explicitly overriding this arg is the least common pattern, but that's now the majority of what our test cases cover.

Instead, I'd recommend:
* revert the test cases to use their old/existing recipe args (aside from some new test cases to prove the new arg)
* adjust the `before/after` of each test case accordingly
* in this PR, have a healthy review/discussion of whether the new `before/after` scenarios are acceptable for existing usage (mainly `ChangeSpringPropertyKey`)

My local clone/IntelliJ is misbehaving today when running failing rewrite tests (`java.lang.SecurityException: class "org.eclipse.jgit.util.FS"'s signer information does not match signer information of other classes in the same package` from the classloader at `org.openrewrite.internal.InMemoryDiffEntry.<init>(InMemoryDiffEntry.java:62)`- something amiss between rewrite's shaded jgit and regular jgit?...), but when I try these (new) test cases:

```java
        @Test
        void foo() {
            rewriteRun(
              spec -> spec.recipe(new ChangePropertyKey("spring.profiles", "spring.config.activate.on-profile", null, null)),
              yaml(
                """
                  spring.profiles: prod
                  """,
                """
                  spring:
                    config:
                      activate:
                        on-profile: prod
                  """
              )
            );
        }

        @Test
        void bar() {
            rewriteRun(
              spec -> spec.recipe(new ChangePropertyKey("spring.profiles", "spring.config.activate.on-profile", null, null)),
              yaml(
                """
                  spring:
                    profiles: prod
                  """,
                """
                  spring:
                    config:
                      activate:
                        on-profile: prod
                  """
              )
            );
        }
```

...the former case fails, having produced:
`spring:  config:    activate:      on-profile: prod` (no newlines).

`rewrite-spring` uses `ChangeSpringPropertyKey` to make a *lot* of changes just like that one, so that would be a considerable regression.

**Comment by joanvr** [comment]
Hi @nmck257, thank you very much for the feedback!
The problem that I see with this recipe, is that it relies on an invalid assumption, that `a.b: c` equals `a: b: c` (with new lines). That's no true for yaml, and when it relies on MergeYaml to do the merges, it usually fails to properly merge the keys, and adding a new "coalesced" key at the end. For most tests that just checks the moving of one single key, this might not be an issue, but when it comes to real cases, we have "new coalesced keys" that do not merge with previously existing properties....
Most of the tests, uses keys with dots (which I do not think are the same as coalesced keys). Because having a key called `files.enabled`, `enable.process.file` or `management.metrics` makes sense semantically, the dot is just as an "space", and does not make sense to always split them. Probably it would be better to use slashes (`-`) instead of dots in a yaml ecosystem, but thats completely out of our scope. Here comes the issue, the usage of the dot (`.`) both as a character in a key and as a key separator in the path to reference a key in the parameters. How we can support outputting some key with dots, when we try to split by dots to merge the keys?

Previously, the behavior was to always "coalesce" the new key, so it will maintain keys with dots joined, but also creates a new "coalesced" key that does not match and cannot be merged properly with existing keys... And as said, for single scenarios checked in a test it seems to work, but fails in a more general context.

I agree that changing all the tests parameters to make them pass is probably not the best idea, because probably we are even removing the original idea around this test, so it might be not needed anymore or we are missing some corner case... But how we can "fix" those tests were there are keys with dots (the default of these tests) without adding this extra parameter? Since this only affects output, because we can sort out the "input" part, maybe the idea of @shanman190 of Json path style with quotes is a good idea, but I'm worried because then we are removing the possibility of having those quotes as a characters in the key (which are valid yaml keys, but probably not used that much?).

And finally, some of the yaml tests (ans usages) of this recipe on rewrite-spring would. need to be changed too... Right now the first test even changes the key `servlet-path` for `servlet.path`, which without any change will result in an "split"... But this will happen no matter the solution we use... on how to keep keys with dots.

**Comment by joanvr** [comment]
> My local clone/IntelliJ is misbehaving today when running failing rewrite tests (`java.lang.SecurityException: class "org.eclipse.jgit.util.FS"'s signer information does not match signer information of other classes in the same package` from the classloader at `org.openrewrite.internal.InMemoryDiffEntry.<init>(InMemoryDiffEntry.java:62)`- something amiss between rewrite's shaded jgit and regular jgit?...), but when I try these (new) test cases:
> 
> ```java
>         @Test
>         void foo() {
>             rewriteRun(
>               spec -> spec.recipe(new ChangePropertyKey("spring.profiles", "spring.config.activate.on-profile", null, null)),
>               yaml(
>                 """
>                   spring.profiles: prod
>                   """,
>                 """
>                   spring:
>                     config:
>                       activate:
>                         on-profile: prod
>                   """
>               )
>             );
>         }
> 
>         @Test
>         void bar() {
>             rewriteRun(
>               spec -> spec.recipe(new ChangePropertyKey("spring.profiles", "spring.config.activate.on-profile", null, null)),
>               yaml(
>                 """
>                   spring:
>                     profiles: prod
>                   """,
>                 """
>                   spring:
>                     config:
>                       activate:
>                         on-profile: prod
>                   """
>               )
>             );
>         }
> ```
> 
> ...the former case fails, having produced: `spring: config: activate: on-profile: prod` (no newlines).
> 
> `rewrite-spring` uses `ChangeSpringPropertyKey` to make a _lot_ of changes just like that one, so that would be a considerable regression.

Fixed this issue with the newlines when the matching key is on the root! Thanks for the feedback! :)

**Comment by nmck257** [comment]
Hey @joanvr,
I think we're approaching this recipe with different design goals. Re:
> The problem that I see with this recipe, is that it relies on an invalid assumption, that a.b: c equals a: b: c (with new lines). That's no true for yaml,

Agreed that this isn't true for general Yaml. But this recipe isn't targeting general Yaml; that's what [`ChangeKey`](https://github.com/openrewrite/rewrite/blob/main/rewrite-yaml/src/main/java/org/openrewrite/yaml/ChangeKey.java) does. This recipe is specifically targeting the Spring Boot config scenario, as described in the comment at the top (this is also implied by the presence of the `relaxedBinding` option, which is explicitly a Spring concern):

https://github.com/openrewrite/rewrite/blob/eda18b5582148608b5dd0d91933ff1519747f260/rewrite-yaml/src/main/java/org/openrewrite/yaml/ChangePropertyKey.java#L40-L47

With that assumption in place, dot-separating vs nesting are both correct and equivalent. Currently, the recipe prefers dot-separating the output, but I think that was an arbitrary choice. If we see benefit in preferring nesting, then that's a fine change.

And it sounds like (based on your comments) that we do see benefit in preferring nesting, for example, in the case where we run `ChangePropertyKey` / `MergeYaml` to produce a bunch similar/overlapping keys.

In that example, if I understand correctly, a preference for dot-separation might produce something like this:
```yml
a:
  b.c.d: qwe
  b.c.e: qwe
  q.z: qwe
  b.g: qwe
```
...whereas a preference for nesting might produce something like this:
```yml
a:
  b:
    c:
      d: qwe
      e: qwe
    g: qwe
  q:
    z: qwe
```
...and we expect that the latter is more-desirable for most users. That's reasonable to me. But my key point is that in the scope of this recipe, both of those outputs are logically correct; just stylistically different.

**Comment by shanman190** [comment]
@joanr, I was thinking about this some more and the quote version in source that you're mentioning, could appear in two different shapes for a single-level property key (unsplit): `a."b.c".d` (omitting the yaml spec's optional key quotes) or `"a.\"b.c\".d"` (using the yaml spec's optional key quotes). Given this, I don't think it would come off confusing to a user to have to escape the inner quotes for keys which appeared in the raw source themselves as they would have to be escaped had the optional, full-key quotes been present.

To me, this seems to match more closely with the heart of the yaml spec as any character can appear in the yaml key, but if it contains a YAML special character them it must be quoted and potentially escaped. Even for the current test cases using `:` as the delimiter, those could appear in raw source keys without any issue if the original source key was quoted (ie. `"a:b"` is a valid yaml key).

**Comment by shanman190** [comment]
I think where the problem comes in is that frameworks like Spring have chosen to interpret `"a.b.c.d": value`, `a: b: c: d: value` (using newlines), and all patterns in between as having equality to each other. This attribute is kinda counter to the yaml spec itself where each of those is truly different and could all appear even in the same document without a collision.

With this said, it makes me wonder if the `rewrite-spring`, etc frameworks really need a specialized yaml visitor that turns it's head to the yaml spec just like their internal configuration parsers do and treats every key path, including those in source, as being their fully separated with new lines equivalent. This would probably simplify some of the issues that we end up seeing here in terms of yaml merging not behaving as a user would expect.
